File: core-render-android/src/main/java/com/tencent/kuikly/core/render/android/performace/KRPerformanceManager.kt
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.kuikly.core.render.android.performace

import com.tencent.kuikly.core.render.android.IKuiklyRenderViewLifecycleCallback
import com.tencent.kuikly.core.render.android.adapter.KuiklyRenderLog
import com.tencent.kuikly.core.render.android.context.KuiklyRenderCoreExecuteModeBase
import com.tencent.kuikly.core.render.android.exception.ErrorReason
import com.tencent.kuikly.core.render.android.performace.frame.KRFrameData
import com.tencent.kuikly.core.render.android.performace.frame.KRFrameMonitor
import com.tencent.kuikly.core.render.android.performace.launch.KRLaunchData
import com.tencent.kuikly.core.render.android.performace.launch.KRLaunchMonitor
import com.tencent.kuikly.core.render.android.performace.memory.KRMemoryData
import com.tencent.kuikly.core.render.android.performace.memory.KRMemoryMonitor

/**
 * 性能监控类型
 */
enum class KRMonitorType {
    LAUNCH, // 启动监控
    FRAME,  // FPS监控
    MEMORY  // 内存监控
}

/**
 * 性能监控管理
 */
class KRPerformanceManager(
    private val pageName: String,
    private val executeMode: KuiklyRenderCoreExecuteModeBase,
    monitorTypes: List<KRMonitorType>
) : IKuiklyRenderViewLifecycleCallback {

    private val monitors = mutableListOf<KRMonitor<*>>()

    private var dataCallback: IKRMonitorCallback? = null

    private var initTimeStamps = 0L
    private var isColdLaunch = false
    private var isPageColdLaunch = false

    companion object {
        private const val TAG = "KRPerformanceManager"
        private var sIsColdLaunch = true
        // 用于记录页面是否首次打开
        private val pageRecords = mutableListOf<String>()
    }

    init {
        monitorTypes.forEach { type ->
            when (type) {
                KRMonitorType.LAUNCH -> monitors.add(KRLaunchMonitor().apply {
                    addListener {
                        dataCallback?.onLaunchResult(it)
                    }
                })
                KRMonitorType.FRAME -> monitors.add(KRFrameMonitor())
                KRMonitorType.MEMORY -> monitors.add(KRMemoryMonitor())
            }
        }
        if (!pageRecords.contains(pageName)) {
            pageRecords.add(pageName)
            isPageColdLaunch = true
        }
    }

    fun <T : KRMonitor<*>> getMonitor(name: String): T? {
        return monitors.find { it.name() == name } as? T
    }

    fun setMonitorCallback(dataCallback: IKRMonitorCallback) {
        this.dataCallback = dataCallback
    }

    override fun onInit() {
        KuiklyRenderLog.d(TAG, "--onInit--")
        initTimeStamps = System.currentTimeMillis()
        if (sIsColdLaunch) {
            isColdLaunch = true
            sIsColdLaunch = false
        }
        monitors.forEach {
            it.onInit()
        }
    }

    override fun onPreloadClassFinish() {
        KuiklyRenderLog.d(TAG, "--onPreloadClassFinish--")
        monitors.forEach {
            it.onPreloadClassFinish()
        }
    }

    override fun onInitCoreStart() {
        KuiklyRenderLog.d(TAG, "--onRenderCoreInitStart--")
        monitors.forEach {
            it.onInitCoreStart()
        }
    }

    override fun onInitCoreFinish() {
        KuiklyRenderLog.d(TAG, "--onRenderCoreInitFinish--")
        monitors.forEach {
            it.onInitCoreFinish()
        }
    }

    override fun onInitContextStart() {
        KuiklyRenderLog.d(TAG, "--onRenderContextInitStart--")
        monitors.forEach {
            it.onInitContextStart()
        }
    }

    override fun onInitContextFinish() {
        KuiklyRenderLog.d(TAG, "--onRenderContextInitFinish--")
        monitors.forEach {
            it.onInitContextFinish()
        }
    }

    override fun onCreateInstanceStart() {
        KuiklyRenderLog.d(TAG, "--onCreatePageStart--")
        monitors.forEach {
            it.onCreateInstanceStart()
        }
    }

    override fun onCreateInstanceFinish() {
        KuiklyRenderLog.d(TAG, "--onRenderPageFinish--")
        monitors.forEach {
            it.onCreateInstanceFinish()
        }
    }

    override fun onResume() {
        KuiklyRenderLog.d(TAG, "--onResume--")
        monitors.forEach {
            it.onResume()
        }
    }

    override fun onFirstFramePaint() {
        KuiklyRenderLog.d(TAG, "--onFirstFramePaint--")
        monitors.forEach {
            it.onFirstFramePaint()
        }
    }

    override fun onPause() {
        KuiklyRenderLog.d(TAG, "--onPause--")
        monitors.forEach {
            it.onPause()
        }
    }

    override fun onDestroy() {
        KuiklyRenderLog.d(TAG, "--onDestroy--")
        val performanceData = getPerformanceData()
        dataCallback?.onResult(performanceData)
        monitors.forEach {
            it.onDestroy()
        }
    }

    override fun onRenderException(throwable: Throwable, errorReason: ErrorReason) {
        KuiklyRenderLog.d(TAG, "--onRenderException--")
        monitors.forEach {
            it.onRenderException(throwable, errorReason)
        }
    }

    /**
     * 收集所有性能数据
     */
    fun getPerformanceData(): KRPerformanceData {
        val launchData = getLaunchData()
        val frameData = getFrameDate()
        val memoryData = getMemoryData()
        val spentTime = System.currentTimeMillis() - initTimeStamps
        return KRPerformanceData(
            pageName,
            spentTime,
            isColdLaunch,
            isPageColdLaunch,
            executeMode.mode,
            launchData,
            frameData,
            memoryData
        )
    }

    /**
     * 获取启动数据
     */
    fun getLaunchData(): KRLaunchData? = getMonitor<KRLaunchMonitor>(KRLaunchMonitor.MONITOR_NAME)?.getMonitorData()

    /**
     * 获取帧数据
     */
    fun getFrameDate(): KRFrameData? = getMonitor<KRFrameMonitor>(KRFrameMonitor.MONITOR_NAME)?.getMonitorData()

    /**
     * 获取内存数据
     */
    fun getMemoryData(): KRMemoryData? = getMonitor<KRMemoryMonitor>(KRMemoryMonitor.MONITOR_NAME)?.getMonitorData()

}

interface IKRMonitorCallback {

    /**
     * 回调启动数据
     */
    fun onLaunchResult(data: KRLaunchData)

    /**
     * 回调所有性能数据
     */
    fun onResult(data: KRPerformanceData)

}

File: core-render-ohos/src/main/cpp/libohos_render/performance/KRPerformanceManager.cpp
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "KRPerformanceManager.h"

#include "libohos_render/performance/KRPerformanceData.h"

bool KRPerformanceManager::cold_launch_flag = true;
std::list<std::string> KRPerformanceManager::page_record_;

KRPerformanceManager::KRPerformanceManager(std::string page_name, const std::shared_ptr<KRRenderExecuteMode> &mode)
    : page_name_(std::move(page_name)), mode_(mode) {
    auto launch_monitor = std::make_shared<KRLaunchMonitor>();
    monitors_[KRLaunchMonitor::kMonitorName] = launch_monitor;
    auto it = std::find(page_record_.begin(), page_record_.end(), page_name_);
    if (it == page_record_.end()) {  //  页面未曾加载过
        is_page_cold_launch = true;
        page_record_.push_front(page_name_);
    }
}

KRPerformanceManager::~KRPerformanceManager() {}

void KRPerformanceManager::SetArkLaunchTime(int64_t launch_time) {  //  ArkTS层页面加载启动事件
    init_time_stamps_ = launch_time;
    if (auto monitor = GetMonitor(KRLaunchMonitor::kMonitorName)) {  //  这个事件只有LaunchMonitor关注
        monitor->SetArkLaunchTime(launch_time);
    }
}

void KRPerformanceManager::OnKRRenderViewInit() {
    if (cold_launch_flag) {
        cold_launch_flag = false;
        is_cold_launch = true;
    }
    for (const auto &monitor : monitors_) {
        monitor.second->OnKRRenderViewInit();
    }
}

void KRPerformanceManager::OnInitCoreStart() {
    for (const auto &monitor : monitors_) {
        monitor.second->OnInitCoreStart();
    }
}
void KRPerformanceManager::OnInitCoreFinish() {
    for (const auto &monitor : monitors_) {
        monitor.second->OnInitCoreFinish();
    }
}
void KRPerformanceManager::OnInitContextStart() {
    for (const auto &monitor : monitors_) {
        monitor.second->OnInitContextStart();
    }
}
void KRPerformanceManager::OnInitContextFinish() {
    for (const auto &monitor : monitors_) {
        monitor.second->OnInitContextFinish();
    }
}
void KRPerformanceManager::OnCreateInstanceStart() {
    for (const auto &monitor : monitors_) {
        monitor.second->OnCreateInstanceStart();
    }
}
void KRPerformanceManager::OnCreateInstanceFinish() {
    for (const auto &monitor : monitors_) {
        monitor.second->OnCreateInstanceFinish();
    }
}
void KRPerformanceManager::OnFirstFramePaint() {
    for (const auto &monitor : monitors_) {
        monitor.second->OnFirstFramePaint();
    }
}

void KRPerformanceManager::OnPageCreateFinish(KRPageCreateTrace &trace) {  //  KuiklyCore回调的页面创建事件
    auto monitor = GetMonitor(KRLaunchMonitor::kMonitorName);
    if (monitor) {
        auto launch_monitor = std::static_pointer_cast<KRLaunchMonitor>(monitor);
        launch_monitor->OnPageCreateFinish(trace);
    }
}
void KRPerformanceManager::OnResume() {}
void KRPerformanceManager::OnPause() {}
void KRPerformanceManager::OnDestroy() {}

std::string KRPerformanceManager::GetPerformanceData() {  //  收集所有性能数据
    auto monitor = GetMonitor(KRLaunchMonitor::kMonitorName);
    if (monitor) {
        auto now = std::chrono::system_clock::
            now();  //  这里用system_clock。原因是KuiklyCore回调的页面创建事件都是epoch_time。
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch());
        auto spent_time = duration.count() - init_time_stamps_;
        int kuikly_core_mode_value = mode_->ModeToCoreValue();
        KRPerformanceData performance =
            KRPerformanceData(page_name_, kuikly_core_mode_value, spent_time, is_cold_launch, is_page_cold_launch,
                              monitor->GetMonitorData());
        return performance.ToJsonString();
    }
    return "{}";
}

std::shared_ptr<KRMonitor> KRPerformanceManager::GetMonitor(std::string monitor_name) {
    if (monitors_.find(KRLaunchMonitor::kMonitorName) != monitors_.end()) {
        return monitors_[monitor_name];
    }
    return nullptr;
}